### PR TITLE
[mlir][ArmSME] Split the Op definition (nfc)

### PR DESCRIPTION
**Pull request by banach-space**
Move the definitions of LLVM intrinsic Ops for ArmSME into a dedicated
file. To facilitate this, the dialect definition together with various
shared definitions are moved to ArmSMEBase.td.

This change will allow us to refactor the ArmSME dialect documentation.
In particular, we will be able to categorise the Ops into "regular"  and
"intrinsic" ops. Also, it will be easier to add some custom
documentation as opposed to relying on auto-generated docs that simply
list the available Ops.

The documentation will be updated in a forthcoming patch. Only
non-functional changes.

### CODE REVIEW

**Comment by MacDue** [comment]
CI does not seem happy with this change :sweat_smile:

**Comment by banach-space** [comment]
> CI does not seem happy with this change 😅

Looks like a missing CMake dependency - I've sent an update that should fix it 🤞🏻

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp [hunk 1]
```diff
@@ -187,10 +187,10 @@ struct LoadTileSliceToArmSMELowering
     auto allActiveMask = rewriter.create<vector::SplatOp>(loc, predTy, one);
 
     auto tileI32 = castTileIDToI32(tile, loc, rewriter);
-    arm_sme::TileSliceLayout layout = loadTileSliceOp.getLayout();
```
Comment (c-rhodes) [inline mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:190]: why has the namespace been removed?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td:1]: I'm wondering if we should rename this to `ArmSMEOps.td`? And the new .td file with the dialect could be `ArmSME.td` or `ArmSMEDialect.td`?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEBase.td [hunk 2]
```diff
@@ -0,0 +1,52 @@
+//===-- ArmSMESMpBase.td - ArmSME dialect definitions -----*- tablegen -*-===//
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEBase.td:1]: filename is wrong 
```suggestion
//===-- ArmSMEBase.td - ArmSME dialect definitions --------*- tablegen -*-===//
```

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEIntrinsicOps.td [hunk 3]
```diff
@@ -0,0 +1,137 @@
+//===-- ArmSMEIntrinsicsOps.td -----------------------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains definitons of the intrinsic Ops for the ArmSME dialect.
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEIntrinsicOps.td:9]: spelling
```suggestion
// This file contains definitions of the intrinsic Ops for the ArmSME dialect.
```

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt [hunk 4]
```diff
@@ -1,12 +1,24 @@
 add_mlir_dialect(ArmSME arm_sme ArmSME)
 add_mlir_doc(ArmSME ArmSME Dialects/ -gen-dialect-doc -dialect=arm_sme)
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:2]: Note that with these changes the docs will be empty as `ArmSME.td` does not include the op definitions. This was the same problem the vector dialect had a little while ago.

Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:2]: #68110 ;-)

Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:2]: No worries then :smile:

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td [hunk 5]
```diff
@@ -1,27 +1,24 @@
...
-// This file defines the ArmSME dialect and contains intrinsic ops to lower to
-// LLVM IR.
+// This file contains the definition of the ArmSME dialect as well as some
+// shared definitions.
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef ARMSME_OPS
-#define ARMSME_OPS
+#ifndef ARMSME_OP_BASE
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td:14]: `ARMSME_DIALECT`?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEIntrinsicOps.td [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains definitions of the intrinsic Ops for the ArmSME dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef ARMSME_INTRINSICS_OPS
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEIntrinsicOps.td:13]: `ARMSME_INTRINSIC_OPS`

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 7]
```diff
@@ -0,0 +1,515 @@
+//===-- ArmSME.td - ArmSME dialect operation definitions ---*- tablegen -*-===//
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td]:
```suggestion
//===-- ArmSMEOps.td - ArmSME dialect operation definitions *- tablegen -*-===//
```

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt [hunk 8]
```diff
@@ -1,12 +1,24 @@
...
-
-mlir_tablegen(ArmSMEEnums.h.inc -gen-enum-decls)
-mlir_tablegen(ArmSMEEnums.cpp.inc -gen-enum-defs)
+# Generate declarations and definitions of ArmSME Ops
+set(LLVM_TARGET_DEFINITIONS ArmSMEOps.td)
+mlir_tablegen(ArmSMEOps.h.inc -gen-op-decls)
+mlir_tablegen(ArmSMEOps.cpp.inc -gen-op-defs)
 mlir_tablegen(ArmSMEAttrDefs.h.inc -gen-attrdef-decls -attrdefs-dialect=arm_sme)
 mlir_tablegen(ArmSMEAttrDefs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=arm_sme)
-add_public_tablegen_target(MLIRArmSMEAttrDefsIncGen)
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:12]: I guess this is part of the `MLIRArmSMEOpsIncGen` target now?

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt [hunk 9]
```diff
@@ -1,12 +1,24 @@
...
+mlir_tablegen(ArmSMEEnums.h.inc -gen-enum-decls)
+mlir_tablegen(ArmSMEEnums.cpp.inc -gen-enum-defs)
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:11]: reordering things isn't helpful when reviewing changes like this

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td (hunk 5 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSME.td:14]: The file is called `ArmSME.td`, so should be plain `ARMSME`. Thanks!

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt (hunk 8 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:12]: Indeed. Everything that's defined in ArmSMEOp.td should lead to one public target (unless there's a good reason not to).

**Review by c-rhodes — APPROVED** [review]
LGTM cheers!

**Review by MacDue — APPROVED** [review]

**Comment by banach-space** [comment]
Thanks for the reviews - I merged this one manually: https://github.com/llvm/llvm-project/commit/8d6d4f8321c740cb3ee5a8a185dc6e3c895b11ca (experimenting with different workflows).

**Comment by MacDue** [comment]
> Thanks for the reviews - I merged this one manually: [8d6d4f8](https://github.com/llvm/llvm-project/commit/8d6d4f8321c740cb3ee5a8a185dc6e3c895b11ca) (experimenting with different workflows).

P.s. If you had put "Resolved #67985" in your commit message this PR would have been automatically closed on commit.

**Pull request closed without merging**